File: ios/Freepass/Freepass/Vault.swift
import Foundation
import SwiftCBOR

enum VaultError : ErrorType {
	case WrongPassword
	case UnknownError
}

struct Vault {
	private static var masterKey : COpaquePointer? = nil
	private static var vaultObj : COpaquePointer? = nil
	private static var entriesKey : COpaquePointer? = nil
	private static var filePath : String? = nil
	private static var isOpen = false

	static func open(path: String, userName: String, password: String) throws {
		masterKey = rusterpassword_gen_master_key(password, userName)
		if masterKey == COpaquePointer.init(nilLiteral: ()) {
			throw VaultError.UnknownError
		}
		let outerKey = freepass_gen_outer_key(masterKey!)
		entriesKey = freepass_gen_entries_key(masterKey!)

		let fm = NSFileManager.defaultManager()
		if fm.isReadableFileAtPath(path) {
			filePath = path
			vaultObj = freepass_open_vault(path, entriesKey!, outerKey)
		} else {
			vaultObj = freepass_new_vault(entriesKey!, outerKey)
		}
		freepass_free_outer_key(outerKey)

		if vaultObj == COpaquePointer.init(nilLiteral: ()) {
			rusterpassword_free_master_key(masterKey!)
			freepass_free_entries_key(entriesKey!)
			throw VaultError.WrongPassword
		}
		isOpen = true
	}

	static func entryNames() -> [String] {
		guard let vaultObj = vaultObj else { return [] }
		let names_iter = freepass_vault_get_entry_names_iterator(vaultObj)
		defer { freepass_free_entry_names_iterator(names_iter) }
		var names = [String]()
		var curr = freepass_entry_names_iterator_next(names_iter)
		while curr != UnsafeMutablePointer.init(nilLiteral: ()) {
			names.append(String.fromCString(curr)!)
			freepass_free_entry_name(curr)
			curr = freepass_entry_names_iterator_next(names_iter)
		}
		return names
	}

	static func getEntry(name: String) -> Entry? {
		guard let vaultObj = vaultObj else { return nil }
		guard let entriesKey = entriesKey else { return nil }
		let cbor = freepass_vault_get_entry_cbor(vaultObj, name)
		defer { freepass_free_entry_cbor(cbor) }
		let bytes = Array(UnsafeBufferPointer(start: cbor.data, count: cbor.len))
		guard let result = try! CBORDecoder(input: bytes).decodeItem() else { return nil }
		return Entry(fromCbor: result)
	}

	static func close() {
		if let masterKey = masterKey {
			rusterpassword_free_master_key(masterKey)
		}
		masterKey = nil
		
		if let entriesKey = entriesKey {
			freepass_free_entries_key(entriesKey)
		}
		
		masterKey = nil
		if let vaultObj = vaultObj {
			freepass_close_vault(vaultObj)
		}
		isOpen = false
	}
}

File: android/app/src/main/kotlin/technology/unrelenting/freepass/Vault.kt
package technology.unrelenting.freepass

import org.bytedeco.javacpp.*
import org.bytedeco.javacpp.annotation.*
import java.io.File

class VaultException(reason: String) : Exception(reason) {}

@Platform(library = "jniVault",
          cinclude = arrayOf("freepass_capi.h"))
object Vault {
	var masterKey: Pointer? = null
	var vaultObj: Pointer? = null
	var entriesKey: Pointer? = null
	var filePath: String? = null
	var isOpen = false

	init {
		//Loader.load() // javacpp's loader tries to compute the full path inside the jar for some reason
		                // and ends up with .../arm/... instead of .../armeabi-v7a/... or .../arm64-v7a/...
		System.loadLibrary("jniVault")
		freepass_init()
	}

	fun open(path: String, username: String, password: String) {
		if (isOpen) {
			throw VaultException("Already open")
		}
		masterKey = rusterpassword_gen_master_key(password, username)
		if (masterKey == null || masterKey!!.isNull) {
			throw VaultException("Got null master key")
		}
		val outerKey = freepass_gen_outer_key(masterKey!!)
		val entriesKey = freepass_gen_entries_key(masterKey!!)

		val file = File(path)
		vaultObj = if (file.exists()) {
			if (!file.canRead()) {
				throw VaultException("Not allowed to read file ${file.absolutePath}")
			}
			freepass_open_vault(file.absolutePath, entriesKey, outerKey)
		} else {
			freepass_new_vault(entriesKey, outerKey)
		}
		freepass_free_outer_key(outerKey)

		if (vaultObj == null || vaultObj!!.isNull) {
			rusterpassword_free_master_key(masterKey!!)
			freepass_free_entries_key(entriesKey)
			throw VaultException("Got null vault")
		}

		filePath = file.absolutePath
		isOpen = true
	}

	fun entryNames(): List<String> {
		if (vaultObj == null) {
			return listOf()
		}
		val iter = freepass_vault_get_entry_names_iterator(vaultObj!!)
		var curr: BytePointer? = freepass_entry_names_iterator_next(iter)
		val result = mutableListOf<String>()
		while (curr != null && !curr.isNull) {
			result.add(curr.string)
			freepass_free_entry_name(curr)
			curr = freepass_entry_names_iterator_next(iter)
		}
		freepass_free_entry_names_iterator(iter)
		return result.toList()
	}

	fun close() {
		if (masterKey != null) {
			rusterpassword_free_master_key(masterKey!!)
		}
		masterKey = null
		if (entriesKey != null) {
			freepass_free_entries_key(entriesKey!!)
		}
		entriesKey = null
		if (vaultObj != null) {
			freepass_close_vault(vaultObj!!)
		}
		vaultObj = null
		isOpen = false
	}

	@JvmStatic external fun rusterpassword_gen_master_key(username: String, password: String): Pointer
	@JvmStatic external fun rusterpassword_free_master_key(@Cast("secstr_t*") master_key: Pointer): Unit

	@JvmStatic external fun freepass_init(): Unit

	@JvmStatic external fun freepass_gen_outer_key(@Cast("secstr_t*") master_key: Pointer): Pointer
	@JvmStatic external fun freepass_gen_entries_key(@Cast("secstr_t*") master_key: Pointer): Pointer
	@JvmStatic external fun freepass_free_outer_key(@Cast("secstr_t*") outer_key: Pointer): Unit
	@JvmStatic external fun freepass_free_entries_key(@Cast("secstr_t*") entries_key: Pointer): Unit

	@JvmStatic external fun freepass_open_vault(path: String, @Cast("secstr_t*") entries_key: Pointer, @Cast("secstr_t*") outer_key: Pointer): Pointer
	@JvmStatic external fun freepass_new_vault(@Cast("secstr_t*") entries_key: Pointer, @Cast("secstr_t*") outer_key: Pointer): Pointer
	@JvmStatic external fun freepass_close_vault(@Cast("vault_t*") vault: Pointer): Unit

	@JvmStatic external fun freepass_vault_get_entry_names_iterator(@Cast("vault_t*") vault: Pointer): Pointer
	@JvmStatic external @Cast("signed char*") fun freepass_entry_names_iterator_next(@Cast("string_iter_t*") iter: Pointer): BytePointer
	@JvmStatic external fun freepass_free_entry_name(@Cast("char*") name: BytePointer): Unit
	@JvmStatic external fun freepass_free_entry_names_iterator(@Cast("string_iter_t*") iter: Pointer): Unit

	@JvmStatic @ByVal external fun freepass_vault_get_entry_cbor(@Cast("vault_t*") vault: Pointer, name: String): vector_t
	@JvmStatic external fun freepass_free_entry_cbor(@ByVal cbor: vector_t)
	@JvmStatic external fun freepass_vault_put_entry_cbor(@Cast("vault_t*") vault: Pointer, name: String, @Cast("uint8_t*") data: BytePointer, @Cast("size_t") len: Int)

	class vector_t : Pointer {
		init { Loader.load() }
		constructor(p: Pointer): super(p) {}
		external fun allocate()
		@Cast("uint8_t*") @MemberGetter external fun data(): BytePointer
		@MemberGetter external fun len(): Int
		@MemberGetter external fun cap(): Int
	}
}